File: testdata/map.go
// Copyright 2011 Jonas mg
//
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// If a copy of the MPL was not distributed with this file, You can obtain one at
// http://mozilla.org/MPL/2.0/.

package main

import "fmt"

var PASS = true
var rating = map[string]float32{"C": 5, "Go": 4.5, "Python": 4.5, "C++": 2}

func builtIn() {
	pass := true

	var m1 map[string]int
	m2 := map[string]int{}
	m3 := make(map[string]int)
	m4 := make(map[string]int, 10)

	tests := []struct {
		msg string
		in  bool
		out bool
	}{
		{"nil m1", m1 == nil, true},
		{"nil m2", m2 == nil, false},
		{"nil m3", m3 == nil, false},
		{"nil m4", m4 == nil, false},
		{"nil m4", m4 != nil, true},

		{"len m1", len(m1) == 0, true},
		{"len m2", len(m2) == 0, true},
		{"len m3", len(m3) == 0, true},
		{"len m4", len(m4) == 0, true},

		{"nil rating", rating != nil, true},
		{"len rating", len(rating) == 4, true},
	}

	for _, t := range tests {
		if t.in != t.out {
			fmt.Printf("\tFAIL: %s => got %v, want %v\n", t.msg, t.in, t.out)
			pass, PASS = false, false
		}
	}
	if pass {
		fmt.Println("\tpass")
	}
}

func declaration() {
	pass := true

	var numbers map[string]float32 // declare a map of strings to ints
	numbers = make(map[string]float32)
	numbers["one"] = 1
	numbers["ten"] = 10
	numbers["trois"] = 3 // trois is "three" in french

	// A map representing the rating given to some programming languages.
	rating1 := map[string]float32{"C": 5, "Go": 4.5, "Python": 4.5, "C++": 2}

	// This is equivalent to writing more verbosely
	rating2 := make(map[string]float32)
	rating2["C"] = 5
	rating2["Go"] = 4.5
	rating2["Python"] = 4.5
	rating2["C++"] = 2

	tests := []struct {
		msg string
		in  float32
		out float32
	}{
		{`numbers["one"]`, numbers["one"], 1},
		{`numbers["ten"]`, numbers["ten"], 10},
		{`numbers["trois"]`, numbers["trois"], 3},

		{`rating["C"]`, rating1["C"], rating2["C"]},
		{`rating["Go"]`, rating1["Go"], rating2["Go"]},
		{`rating["Python"]`, rating1["Python"], rating2["Python"]},
		{`rating["C++"]`, rating1["C++"], rating2["C++"]},
	}

	for _, t := range tests {
		if t.in != t.out {
			fmt.Printf("\tFAIL: %s => got %v, want %v\n", t.msg, t.in, t.out)
			pass, PASS = false, false
		}
	}
	if pass {
		fmt.Println("\tpass")
	}
}

func reference() {
	m := make(map[string]string)
	m["Hello"] = "Bonjour"

	m1 := m
	m1["Hello"] = "Salut" // Now: m["Hello"] == "Salut"

	if m["Hello"] == m1["Hello"] {
		fmt.Println("\tpass")
	} else {
		fmt.Printf("\tFAIL: m[\"Hello\"] => got %v, want %v\n", m["Hello"], m1["Hello"])
		PASS = false
	}
}

func keyNoExistent() {
	pass := true

	csharp_rating := rating["C#"]
	csharp_rating2, found := rating["C#"]

	multiDim := map[int]map[int]float32{1: {1: 1.1}, 2: {2: 2.2}}
	k_multiDim := multiDim[1][2]

	tests := []struct {
		msg string
		in  float32
		out float32
	}{
		{"csharp_rating", csharp_rating, 0.00},
		{"csharp_rating2", csharp_rating2, 0},
		{"k_multiDim", k_multiDim, 0},
	}

	for _, t := range tests {
		if t.in != t.out {
			fmt.Printf("\tFAIL: %s => got %v, want %v\n", t.msg, t.in, t.out)
			pass, PASS = false, false
		}
	}
	if found {
		fmt.Printf("\tFAIL: using comma => got %v, want %v\n", found, !found)
		pass, PASS = false, false
	}
	if pass {
		fmt.Println("\tpass")
	}
}

func deleteKey() {
	pass := true

	delete(rating, "C++")
	_, found := rating["C++"]

	if found {
		fmt.Printf("\tFAIL: got %v, want %v\n", found, !found)
		pass, PASS = false, false
	}
	if pass {
		fmt.Println("\tpass")
	}
}

func _range() {
	pass := true

	for key, value := range rating {
		switch key {
		case "C":
			if value != 5 {
				fmt.Printf("\tFAIL: %s => got %v, want 5\n", key, value)
				pass, PASS = false, false
			}
		case "Go":
			if value != 4.5 {
				fmt.Printf("\tFAIL: %s => got %v, want 4.5\n", key, value)
				pass, PASS = false, false
			}
		case "Python":
			if value != 4.5 {
				fmt.Printf("\tFAIL: %s => got %v, want 4.5\n", key, value)
				pass, PASS = false, false
			}
		default:
			fmt.Printf("\tFAIL: %s => no expected\n", key)
			pass, PASS = false, false
		}
	}

	// Omit the value.
	for key := range rating {
		if key != "C" && key != "Go" && key != "Python" {
			fmt.Printf("\tFAIL: key %q no expected\n", key)
			pass, PASS = false, false
		}
	}

	if pass {
		fmt.Println("\tpass")
	}
}

func blankIdInRange() {
	pass := true

	// Return the biggest value in a slice of ints.
	Max := func(slice []int) int {
		max := slice[0] // The first element is the max for now.
		for _, value := range slice {
			if value > max { // We found a bigger value in our slice.
				max = value
			}
		}
		return max
	}

	var slice []int
	// Declare three arrays of different sizes, to test the function Max.
	A1 := [10]int{1, 2, 3, 4, 5, 6, 7, 8, 9}
	A2 := [4]int{1, 2, 3, 4}
	A3 := [1]int{1}

	slice = A1[:] // Take all A1 elements.
	if Max(slice) != 9 {
		fmt.Printf("\tFAIL: A1 => got %v, want 9\n", Max(slice))
		pass, PASS = false, false
	}
	slice = A2[:]
	if Max(slice) != 4 {
		fmt.Printf("\tFAIL: A2 => got %v, want 4\n", Max(slice))
		pass, PASS = false, false
	}
	slice = A3[:]
	if Max(slice) != 1 {
		fmt.Printf("\tFAIL: A3 => got %v, want 1\n", Max(slice))
		pass, PASS = false, false
	}

	if pass {
		fmt.Println("\tpass")
	}
}

func main() {
	fmt.Print("\n\n== Maps\n\n")

	fmt.Println("=== RUN builtIn")
	builtIn()
	fmt.Println("=== RUN declaration")
	declaration()
	fmt.Println("=== RUN reference")
	reference()
	fmt.Println("=== RUN keyNoExistent")
	keyNoExistent()
	fmt.Println("=== RUN deleteKey")
	deleteKey()
	fmt.Println("=== RUN range")
	_range()
	fmt.Println("=== RUN blankIdInRange")
	blankIdInRange()

	if PASS {
		fmt.Println("PASS")
	} else {
		fmt.Println("FAIL")
		print("Fail: Maps")
	}
}

File: testdata/map.js










var PASS = true;
var rating = g.Map(0, {"C": 5, "Go": 4.5, "Python": 4.5, "C++": 2});

function builtIn() {
	var pass = true;

	var m1 = g.Map(0);
	var m2 = g.Map(0, {});
	var m3 = g.Map(0, {});
	var m4 = g.Map(0, {});

	var _ = function(msg, in_, out) { return {
		msg: msg,
		in_: in_,
		out: out
	};}; var tests = [
		_("nil m1", m1.v == undefined, true),
		_("nil m2", m2.v == undefined, false),
		_("nil m3", m3.v == undefined, false),
		_("nil m4", m4.v == undefined, false),
		_("nil m4", m4.v != undefined, true),

		_("len m1", m1.len() == 0, true),
		_("len m2", m2.len() == 0, true),
		_("len m3", m3.len() == 0, true),
		_("len m4", m4.len() == 0, true),

		_("nil rating", rating.v != undefined, true),
		_("len rating", rating.len() == 4, true)
	];

	var t; for (var _ in tests) { t = tests[_];
		if (JSON.stringify(t.in_) != JSON.stringify(t.out)) {
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: " + t.msg + " => got " + t.in_ + ", want " + t.out + "<br>");
			pass = false, PASS = false;
		}
	}
	if (pass) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	}
}

function declaration() {
	var pass = true;

	var numbers = g.Map(0);
	numbers = g.Map(0, {});
	numbers.v["one"] = 1;
	numbers.v["ten"] = 10;
	numbers.v["trois"] = 3;


	var rating1 = g.Map(0, {"C": 5, "Go": 4.5, "Python": 4.5, "C++": 2});


	var rating2 = g.Map(0, {});
	rating2.v["C"] = 5;
	rating2.v["Go"] = 4.5;
	rating2.v["Python"] = 4.5;
	rating2.v["C++"] = 2;

	var _ = function(msg, in_, out) { return {
		msg: msg,
		in_: in_,
		out: out
	};}; var tests = [
		_("numbers[\"one\"]", numbers.get("one")[0], 1),
		_("numbers[\"ten\"]", numbers.get("ten")[0], 10),
		_("numbers[\"trois\"]", numbers.get("trois")[0], 3),

		_("rating[\"C\"]", rating1.get("C")[0], rating2.get("C")[0]),
		_("rating[\"Go\"]", rating1.get("Go")[0], rating2.get("Go")[0]),
		_("rating[\"Python\"]", rating1.get("Python")[0], rating2.get("Python")[0]),
		_("rating[\"C++\"]", rating1.get("C++")[0], rating2.get("C++")[0])
	];

	var t; for (var _ in tests) { t = tests[_];
		if (JSON.stringify(t.in_) != JSON.stringify(t.out)) {
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: " + t.msg + " => got " + t.in_ + ", want " + t.out + "<br>");
			pass = false, PASS = false;
		}
	}
	if (pass) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	}
}

function reference() {
	var m = g.Map("", {});
	m.v["Hello"] = "Bonjour";

	var m1 = m;
	m1.v["Hello"] = "Salut";

	if (JSON.stringify(m.get("Hello")[0]) == JSON.stringify(m1.get("Hello")[0])) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	} else {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: m[\"Hello\"] => got " + m.get("Hello")[0] + ", want " + m1.get("Hello")[0] + "<br>");
		PASS = false;
	}
}

function keyNoExistent() {
	var pass = true;

	var csharp_rating = rating.get("C#")[0];
	var _ = rating.get("C#"), csharp_rating2 = _[0], found = _[1];

	var multiDim = g.Map(0, {1: {1: 1.1}, 2: {2: 2.2}});
	var k_multiDim = multiDim.get(1,2)[0];

	var _ = function(msg, in_, out) { return {
		msg: msg,
		in_: in_,
		out: out
	};}; var tests = [
		_("csharp_rating", csharp_rating, 0.00),
		_("csharp_rating2", csharp_rating2, 0),
		_("k_multiDim", k_multiDim, 0)
	];

	var t; for (var _ in tests) { t = tests[_];
		if (JSON.stringify(t.in_) != JSON.stringify(t.out)) {
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: " + t.msg + " => got " + t.in_ + ", want " + t.out + "<br>");
			pass = false, PASS = false;
		}
	}
	if (found) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: using comma => got " + found + ", want " + !found + "<br>");
		pass = false, PASS = false;
	}
	if (pass) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	}
}

function deleteKey() {
	var pass = true;

	delete rating.v["C++"];
	var found = rating.get("C++")[1];

	if (found) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: got " + found + ", want " + !found + "<br>");
		pass = false, PASS = false;
	}
	if (pass) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	}
}

function _range() {
	var pass = true;

	var value; for (var key in rating.v) { value = rating.get(key)[0];
		switch (key) {
		case "C":
			if (value != 5) {
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: " + key + " => got " + value + ", want 5<br>");
			pass = false, PASS = false;
		} break;
		case "Go":
			if (value != 4.5) {
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: " + key + " => got " + value + ", want 4.5<br>");
			pass = false, PASS = false;
		} break;
		case "Python":
			if (value != 4.5) {
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: " + key + " => got " + value + ", want 4.5<br>");
			pass = false, PASS = false;
		} break;
		default:
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: " + key + " => no expected<br>");
			pass = false, PASS = false;
		}
	}


	for (var key in rating.v) {
		if (key != "C" && key != "Go" && key != "Python") {
			document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: key " + key + " no expected<br>");
			pass = false, PASS = false;
		}
	}

	if (pass) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	}
}

function blankIdInRange() {
	var pass = true;


	var Max = function(slice) {
		var max = slice.get()[0];
		var value; for (var _ in slice.get()) { value = slice.get()[_];
			if (value > max) {
				max = value;
			}
		}
		return max;
	};

	var slice = g.MkSlice();

	var A1 = g.MkArray([10], 0, [1, 2, 3, 4, 5, 6, 7, 8, 9]);
	var A2 = g.MkArray([4], 0, [1, 2, 3, 4]);
	var A3 = g.MkArray([1], 0, [1]);

	slice = g.SliceFrom(A1, 0);
	if (Max(slice) != 9) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: A1 => got " + Max(slice) + ", want 9<br>");
		pass = false, PASS = false;
	}
	slice = g.SliceFrom(A2, 0);
	if (Max(slice) != 4) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: A2 => got " + Max(slice) + ", want 4<br>");
		pass = false, PASS = false;
	}
	slice = g.SliceFrom(A3, 0);
	if (Max(slice) != 1) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;FAIL: A3 => got " + Max(slice) + ", want 1<br>");
		pass = false, PASS = false;
	}

	if (pass) {
		document.write("&nbsp;&nbsp;&nbsp;&nbsp;pass<br>");
	}
}

function main() {
	document.write("<br><br>== Maps<br><br>");

	document.write("=== RUN builtIn<br>");
	builtIn();
	document.write("=== RUN declaration<br>");
	declaration();
	document.write("=== RUN reference<br>");
	reference();
	document.write("=== RUN keyNoExistent<br>");
	keyNoExistent();
	document.write("=== RUN deleteKey<br>");
	deleteKey();
	document.write("=== RUN range<br>");
	_range();
	document.write("=== RUN blankIdInRange<br>");
	blankIdInRange();

	if (PASS) {
		document.write("PASS<br>");
	} else {
		document.write("FAIL<br>");
		alert("Fail: Maps");
	}
} main();
/* Generated by Go2js (github.com/tredoe/go2js) */
